src/node_api.h

```diff
@@ -0,0 +1,483 @@
+  static void fn(void) __attribute__((constructor)); \
+  static void fn(void)
+#endif
+
+#define NAPI_MODULE_X(modname, regfunc, priv, flags)                  \
+  extern "C" {                                                        \
+    static napi_module _module =                                      \
+    {                                                                 \
+      NAPI_MODULE_VERSION,                                            \
+      flags,                                                          \
+      __FILE__,                                                       \
+      regfunc,                                                        \
+      #modname,                                                       \
+      priv,                                                           \
+      {0},                                                            \
+    };                                                                \
+    NAPI_C_CTOR(_register_ ## modname) {                              \
+      napi_module_register(&_module);                                 \
+    }                                                                 \
+  }
+
+#define NAPI_MODULE(modname, regfunc) \
+  NAPI_MODULE_X(modname, regfunc, NULL, 0)
+
+extern "C" {
+
+NAPI_EXTERN void napi_module_register(napi_module* mod);
+
+NAPI_EXTERN const napi_extended_error_info* napi_get_last_error_info();
+
+// Getters for defined singletons
+NAPI_EXTERN napi_status napi_get_undefined(napi_env env, napi_value* result);
+NAPI_EXTERN napi_status napi_get_null(napi_env env, napi_value* result);
+NAPI_EXTERN napi_status napi_get_false(napi_env env, napi_value* result);
+NAPI_EXTERN napi_status napi_get_true(napi_env env, napi_value* result);
+NAPI_EXTERN napi_status napi_get_global(napi_env env, napi_value* result);
+
+// Methods to create Primitive types/Objects
+NAPI_EXTERN napi_status napi_create_object(napi_env env, napi_value* result);
+NAPI_EXTERN napi_status napi_create_array(napi_env env, napi_value* result);
+NAPI_EXTERN napi_status napi_create_array_with_length(napi_env env,
+                                                      int length,
+                                                      napi_value* result);
+NAPI_EXTERN napi_status napi_create_number(napi_env env,
+                                           double val,
+                                           napi_value* result);
+NAPI_EXTERN napi_status napi_create_string_utf8(napi_env env,
+                                                const char* s,
+                                                int length,
+                                                napi_value* result);
+NAPI_EXTERN napi_status napi_create_string_utf16(napi_env env,
+                                                 const char16_t* s,
+                                                 int length,
+                                                 napi_value* result);
```

Comment:
Good feedback, we should consider adding it. I opened https://github.com/nodejs/abi-stable-node/issues/179 to track this.

Note *missing* APIs should probably not block this PR unless it is something critical. The goal for this PR is to establish an initial set of APIs that provide enough functionality for native module developers to begin experimenting and give us feedback. We have plans to expand on these APIs in the coming months, before the feature graduates out of experimental status.